File: server/mail.mjs
// ─────────────────────────────────────────────────────────────────────
// Entregar email.
//
// Dois caminhos, e o mesmo contrato para quem chama: `deliver` devolve
// true se saiu, false se não. Acrescentar um terceiro fornecedor é
// acrescentar um ramo aqui e uma linha em config.mjs — mais nada no
// resto do servidor precisa de saber qual é.
// ─────────────────────────────────────────────────────────────────────
import { MAIL } from './config.mjs';

const TIMEOUT = 10000;

async function viaResend(msg) {
  const res = await fetch('https://api.resend.com/emails', {
    method: 'POST',
    headers: { Authorization: 'Bearer ' + MAIL.key, 'Content-Type': 'application/json' },
    body: JSON.stringify({
      from: MAIL.from,
      to: [msg.to || MAIL.to],
      reply_to: msg.replyTo || undefined,
      subject: msg.subject,
      text: msg.text,
    }),
    signal: AbortSignal.timeout(TIMEOUT),
  });
  return res.ok;
}

async function viaWebhook(msg) {
  const res = await fetch(MAIL.webhook, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      to: msg.to || MAIL.to,
      from: msg.replyTo || MAIL.from,
      subject: msg.subject,
      message: msg.text,
      source: 'heldergoncalves.io',
    }),
    signal: AbortSignal.timeout(TIMEOUT),
  });
  return res.ok;
}

/**
 * Envia um email.
 * @param {{to?: string, replyTo?: string, subject: string, text: string}} msg
 * @returns {Promise<boolean>} true se o fornecedor aceitou.
 */
export function sendMail(msg) {
  if (MAIL.provider === 'resend') return viaResend(msg);
  return viaWebhook(msg);
}

/** O caso mais comum: alguém escreve ao Hélder a partir do site. */
export function deliverToOwner({ from, subject, message }) {
  return sendMail({
    replyTo: from,
    subject: '[site] ' + subject,
    text: 'Mensagem de heldergoncalves.io\n\nDe: ' + from + '\nAssunto: ' + subject + '\n\n' + message + '\n',
  });
}

File: api/app/mail.py
# ─────────────────────────────────────────────────────────────────────
# Entregar email.
#
# Dois caminhos, e o mesmo contrato para quem chama: `send_mail` devolve
# True se saiu, False se não. Acrescentar um terceiro fornecedor é
# acrescentar um ramo aqui e uma linha em config.py — mais nada no resto
# da API precisa de saber qual é.
# ─────────────────────────────────────────────────────────────────────
from typing import Optional

import httpx

from app.config import MAIL

TIMEOUT = 10.0


async def _via_resend(to: str, reply_to: Optional[str], subject: str, text: str) -> bool:
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        res = await client.post(
            "https://api.resend.com/emails",
            headers={"Authorization": f"Bearer {MAIL.key}", "Content-Type": "application/json"},
            json={
                "from": MAIL.from_,
                "to": [to],
                **({"reply_to": reply_to} if reply_to else {}),
                "subject": subject,
                "text": text,
            },
        )
    return res.status_code < 300


async def _via_webhook(to: str, reply_to: Optional[str], subject: str, text: str) -> bool:
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        res = await client.post(
            MAIL.webhook,
            json={
                "to": to,
                "from": reply_to or MAIL.from_,
                "subject": subject,
                "message": text,
                "source": "heldergoncalves.io",
            },
        )
    return res.status_code < 300


async def send_mail(subject: str, text: str, to: Optional[str] = None, reply_to: Optional[str] = None) -> bool:
    dest = to or MAIL.to
    try:
        if MAIL.provider == "resend":
            return await _via_resend(dest, reply_to, subject, text)
        return await _via_webhook(dest, reply_to, subject, text)
    except httpx.HTTPError:
        return False


async def deliver_to_owner(from_: str, subject: str, message: str) -> bool:
    """O caso mais comum: alguém escreve ao Hélder a partir do site."""
    return await send_mail(
        subject="[site] " + subject,
        text=f"Mensagem de heldergoncalves.io\n\nDe: {from_}\nAssunto: {subject}\n\n{message}\n",
        reply_to=from_,
    )
